providers/okta/application.go

```diff
@@ -0,0 +1,252 @@
+package okta
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"strings"
+
+	"golang.org/x/net/html"
+	"golang.org/x/net/publicsuffix"
+)
+
+// StateToken is a token extracted from the body of an Okta response.
+//
+// It is used to track a single session.
+type StateToken string
+
+type SignedToken struct {
+	Tx  string
+	App string
+}
+
+// ConcatenateAuthSignature applies the auth signature to the token, returning a completed challenge response that can be submitted to Okta.
+func (tok SignedToken) ConcatenateAuthSignature(authSignature string) string {
+	return strings.Join([]string{authSignature, tok.App}, ":")
+}
+
+// ParseSignedToken extracts a transaction and application signature from a signed token from Okta.
+//
+// A signed token is returned from Okta when initiating MFA with Duo.
+func ParseSignedToken(token string) (SignedToken, error) {
+	parts := strings.Split(token, ":")
+	if len(parts) != 2 {
+		return SignedToken{}, errors.New("wrong number of parts in signed token")
+	}
+
+	return SignedToken{Tx: parts[0], App: parts[1]}, nil
+}
+
+type MultiFactorUpgradeMethod interface {
+	Upgrade(ctx context.Context, client *http.Client) (StateToken, error)
+}
+
+type AuthenticatorEnrollment struct {
+	Key            string
+	ContextualData struct {
+		Host        string
+		SignedToken string
+	}
+}
+
+type Remediation struct {
+	Name   string
+	Type   string
+	Href   string
+	Method string
+}
+
+type IdentifyResponse struct {
+	StateHandle                    string
+	CurrentAuthenticatorEnrollment struct {
+		Type  string
+		Value AuthenticatorEnrollment
+	}
+
+	Remediation struct {
+		Type  string
+		Value []Remediation
+	}
+}
+
+func DetermineUpgradePath(resp IdentifyResponse) (MultiFactorUpgradeMethod, bool) {
+	// Two iterations is fine, this is a small list
+	// Prefer Duo OIDC flow
+	for _, rem := range resp.Remediation.Value {
+		if rem.Type == "OIDC" {
+			return DuoFrameless{remediation: rem}, true
+		}
+	}
+
+	for _, rem := range resp.Remediation.Value {
+		// This is intentionally 'Name' as Type is missing for cases where this is true.
+		if rem.Name == "challenge-authenticator" && resp.CurrentAuthenticatorEnrollment.Value.Key == "duo" {
+			host := resp.CurrentAuthenticatorEnrollment.Value.ContextualData.Host
+			tok, err := ParseSignedToken(resp.CurrentAuthenticatorEnrollment.Value.ContextualData.SignedToken)
+			if err != nil {
```

Comment:
If we break the loop because of an error, is there any chance the flow will continue normally? Or should we log and exit?

Reply:
This function has the signature

```go
func DetermineUpgradePath(resp IdentifyResponse) (MultiFactorUpgradeMethod, bool)
```

Where the second bool is a "was present" check, much like map reading: `val, ok := map[key}`. The caller is responsible for formulating the correct error:

```go

	method, ok := DetermineUpgradePath(identifyResponse)
	if !ok {
		return nil, errors.New("could determine 2fa upgrade path - users probably does not have a supported device")
	}
```

which in turn gets passed up the chain and logged.